File: RaftKotlin/raft-node/src/main/kotlin/raft/node/RaftNodeTransport.kt
package raft.node

import raft.types.peer.*
import raft.types.Peer

/**
 * Interface for transport layer for Raft peers.
 * This allows for different transport layers to be used.
 */
interface RaftPeerTransport {
    /**
     * Sends an AppendEntries message to the specified peer.
     * - Parameters:
     *   - request: The AppendEntries message to send.
     *   - peer: The peer to send the message to.
     * - Returns: The response from the peer.
     * - Throws: An error if the message could not be sent.
     */
    @Throws(Exception::class)
    suspend fun appendEntries(
        request: AppendEntriesRequest,
        peer: Peer
    ): AppendEntriesResponse

    /**
     * Sends a RequestVote message to the specified peer.
     * - Parameters:
     *   - request: The RequestVote message to send.
     *   - peer: The peer to send the message to.
     * - Returns: The response from the peer.
     * - Throws: An error if the message could not be sent.
     */
    @Throws(Exception::class)
    suspend fun requestVote(
        request: RequestVoteRequest,
        peer: Peer
    ): RequestVoteResponse

    /**
     * Sends an InstallSnapshot message to the specified peer.
     * - Parameters:
     *   - request: The InstallSnapshot message to send.
     *   - peer: The peer to send the message to.
     * - Returns: The response from the peer.
     * - Throws: An error if the message could not be sent.
     */
    @Throws(Exception::class)
    suspend fun installSnapshot(
        request: InstallSnapshotRequest,
        peer: Peer
    ): InstallSnapshotResponse
}

File: RaftSwift/Sources/RaftCore/Node/RaftNodeTransport.swift
/// Abstraction over transport layer for Raft peers.
/// This allows for different transport layers to be used, e.g. GRPC or Swift Distributed Actors.
public protocol RaftNodeTransport: Sendable {
    /// Sends an AppendEntries message to the specified peer.
    /// - Parameters:
    ///   - request: The AppendEntries message to send.
    ///   - peer: The peer to send the message to.
    ///   - isolation: The isolation to use for the message.
    /// - Returns: The response from the peer.
    /// - Throws: An error if the message could not be sent.
    func appendEntries(
        _ request: AppendEntriesRequest,
        to peer: Peer,
        isolation: isolated (any Actor),
    ) async throws -> AppendEntriesResponse

    /// Sends a RequestVote message to the specified peer.
    /// - Parameters:
    ///   - request: The RequestVote message to send.
    ///   - peer: The peer to send the message to.
    ///   - isolation: The isolation to use for the message.
    /// - Returns: The response from the peer.
    /// - Throws: An error if the message could not be sent.
    func requestVote(
        _ request: RequestVoteRequest,
        to peer: Peer,
        isolation: isolated (any Actor),
    ) async throws -> RequestVoteResponse

    /// Sends an InstallSnapshot message to the specified peer.
    /// - Parameters:
    ///   - request: The InstallSnapshot message to send.
    ///   - peer: The peer to send the message to.
    ///   - isolation: The isolation to use for the message.
    /// - Returns: The response from the peer.
    /// - Throws: An error if the message could not be sent.
    func installSnapshot(
        _ request: InstallSnapshotRequest,
        on peer: Peer,
        isolation: isolated (any Actor),
    ) async throws -> InstallSnapshotResponse
}
